File: fixtures/scale/dart/medium/lib/eval/function_registry.dart
import '../errors/eval_exception.dart';

typedef UnaryFn = double Function(double);
typedef AggregateFn = double Function(List<double>);

class FunctionRegistry {
  final Map<String, UnaryFn> _unary = {};
  final Map<String, AggregateFn> _aggregate = {};

  void addUnary(String name, UnaryFn fn) {
    _unary[name] = fn;
  }

  void addAggregate(String name, AggregateFn fn) {
    _aggregate[name] = fn;
  }

  double invoke(String name, List<double> args) {
    if (args.length == 1 && _unary.containsKey(name)) return _unary[name]!(args[0]);
    final fn = _aggregate[name];
    if (fn == null) throw EvalException('unknown function $name/${args.length}');
    return fn(args);
  }

  bool knows(String name) => _unary.containsKey(name) || _aggregate.containsKey(name);
}

File: fixtures/scale/java/medium/src/calc/eval/FunctionRegistry.java
package calc.eval;

import java.util.HashMap;
import java.util.Map;
import java.util.function.DoubleUnaryOperator;

import calc.errors.EvalException;

public class FunctionRegistry {
    private final Map<String, DoubleUnaryOperator> unary = new HashMap<>();
    private final Map<String, AggregateFunction> aggregate = new HashMap<>();

    public void addUnary(String name, DoubleUnaryOperator fn) {
        unary.put(name, fn);
    }

    public void addAggregate(String name, AggregateFunction fn) {
        aggregate.put(name, fn);
    }

    public double invoke(String name, double[] args) {
        if (args.length == 1 && unary.containsKey(name)) {
            return unary.get(name).applyAsDouble(args[0]);
        }
        AggregateFunction fn = aggregate.get(name);
        if (fn == null) {
            throw new EvalException("unknown function " + name + "/" + args.length);
        }
        return fn.applyAll(args);
    }

    public boolean knows(String name) {
        return unary.containsKey(name) || aggregate.containsKey(name);
    }
}
